File: Sean的leetcode/python/Populating Next Right Pointers in Each Node II.py
###
# BFS
# Time Complexity: O(n)
# Space Complexity: O(n)
###
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=0, left=None, right=None, next=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
class Solution(object):
    def connect(self, root):
        """
        :type root: Node
        :rtype: Node
        """
        if not root:
            return None
        q = collections.deque([root])
        while q:
            pre = None
            level_size = len(q)
            for i in range(level_size):
                cur = q.popleft()
                if pre:
                    pre.next = cur
                if cur.left:
                    q.append(cur.left)
                if cur.right:
                    q.append(cur.right)
                pre = cur
        return root
###
# Follow up: constant space
# Time Complexity: O(n)
# Space Complexity: O(1)
###
###
# Definition for binary tree with next pointer.
# class TreeLinkNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#         self.next = None
class Solution:
    # @param root, a tree link node
    # @return nothing
    def connect(self, root):
        if not root:
            return None
        next_h = pre = None
        cur = root
        while cur:
            if cur.left:
                if not next_h:
                    next_h = cur.left
                else:
                    pre.next = cur.left
                pre = cur.left
            if cur.right:
                if not next_h:
                    next_h = cur.right
                else:
                    pre.next = cur.right
                pre = cur.right
            if cur.next:
                cur = cur.next
            else:
                cur = next_h
                next_h = pre = None

File: MyLeetCode/Leetcode/Populating Next Right Pointers in Each Node II.cpp
//1. refer to http://jane4532.blogspot.tw/2013/07/populate-pointers-to-each-node.html
//2. use BFS concept.
/**
 * Definition for binary tree with next pointer.
 * struct TreeLinkNode {
 *  int val;
 *  TreeLinkNode *left, *right, *next;
 *  TreeLinkNode(int x) : val(x), left(NULL), right(NULL), next(NULL) {}
 * };
 */
class Solution {
public:
    void connect(TreeLinkNode *root) {
        TreeLinkNode *curr = root, *next = 0;
        TreeLinkNode *pre = 0;

        while (curr)
        {
            for (; curr; curr = curr->next)
            {
                if (!curr->left && !curr->right)
                    continue;

                if (curr->left)
                    curr->left->next = curr->right;

                if (pre)
                    pre->next = curr->left ? curr->left : curr->right;

                if (curr->left || curr->right)
                    pre = curr->right ? curr->right : curr->left;

                if (!next)
                    next = curr->left ? curr->left : curr->right;
            }
            curr = next;
            next = pre = 0;
        }
    }
};